packages/discovery-react-components/src/components/DocumentPreview/components/PdfViewerHighlight/utils/common/documentUtils.ts

```diff
@@ -0,0 +1,64 @@
+import { TextMappings } from 'components/DocumentPreview/types';
+import { getTextMappings } from 'components/DocumentPreview/utils/documentData';
+import { QueryResult } from 'ibm-watson/discovery/v2';
+import { processDoc, ProcessedDoc } from 'utils/document';
+import { Location } from 'utils/document/processDoc';
+import { DocumentFields, TextSpan } from '../../types';
+
+/**
+ * Get value of the specified field from a search result document
+ *
+ * @param document search result document
+ * @param field field name
+ * @param index field index. 0 by default
+ * @param span (optional) span on the field value to return. Returns entire the field value by default
+ * @returns text
+ */
+export function getDocFieldValue(
+  document: DocumentFields,
+  field: string,
+  index?: number,
+  span?: Location | TextSpan
+): string | undefined {
+  let fieldText: string | undefined;
+
+  const documentFieldArray = document[field];
+  if (!Array.isArray(documentFieldArray) && !index) {
+    fieldText = documentFieldArray;
+  } else {
+    fieldText = documentFieldArray?.[index ?? 0];
+  }
+  if (!fieldText || !span) {
+    return fieldText;
+  }
+
+  if (Array.isArray(span)) {
+    return fieldText.substring(span[0], span[1]);
+  } else {
+    return fieldText.substring(span.begin, span.end);
+  }
+}
+
+export type ExtractedDocumentInfo = {
+  processedDoc: ProcessedDoc;
+  textMappings?: TextMappings;
+};
+
+/**
+ * Extract bboxes and text_mappings from a search result document
+ */
+export async function extractDocumentInfo(document: QueryResult): Promise<ExtractedDocumentInfo> {
+  const docHtml = document.html;
+  const textMappings = getTextMappings(document) ?? undefined;
+
+  const processedDoc = await processDoc(
```

Comment:
I have't fully mapped out the flow of the code, but is there a possibility of `processDoc` getting called multiple times for the same document? That would be expensive.

Reply:
This is expected to be called once for a document.